### PR TITLE
implement `DataType::try_form(&str)`

### LINKED ISSUE #3821: Implement `FromStr` for DataType / Parse DataType description
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
As part of implementing https://github.com/apache/arrow-datafusion/issues/5016 in DataFusion I needed some way to convert from a string passed in by the user to a `DataType`. 

Since we already had a function `arrow_typeof` that provides a useful human readable type name by calling `DataType::to_string()` I wanted the opposite: A way to take the output of `DataType::to_string()` and make a `DataType`

**Describe the solution you'd like**

I think having `FromStr` implementation that matches the  https://docs.rs/arrow/34.0.0/arrow/datatypes/enum.DataType.html#impl-Display-for-DataType implementation for DataType would be very nice.

Example usage:

```rust
let data_type = DataType:Int32;

// use (existing) Display impl to get a string representaiton
let data_type_string = data_type.to_string();
assert_eq!(data_type_string, "Int32");

// use proposed FromStr impl to get the DataType back
let parsed_datatype: DataType = data_type_string.parse()?
assert_eq!(parsed_datatype, DataType::Int32);
```

**Describe alternatives you've considered**
@tustvold  pointed out that there is already a way to encode data types in String for the IPC format https://arrow.apache.org/docs/format/CDataInterface.html#data-type-description-format-strings

While this format is (designed to be) easy to parse by computers, I don't think it is easy to parse by Humans (quick quiz, what type does `tdD` represent?)

**Additional context**
See https://github.com/apache/arrow-datafusion/pull/5166#discussion_r1129692579

We can probably lift the implementation from https://github.com/apache/arrow-datafusion/pull/5166 into Arrow-rs 

The implementation in  https://github.com/apache/arrow-datafusion/pull/5166 currently does not cover:
- [ ] Parsing Timezones in Timestamp types (@waitingkuo  expressed interest in this)
- [ ] Parsing Struct, Union, Map or List types (though it does handle `Dictionary`)

### PR DESCRIPTION
**Pull request by samuelcolvin**
# Which issue does this PR close?

Closes #3821

# Rationale for this change
 
I would like to use this functionality, currently it's private in datafusion.

# What changes are included in this PR?

`DataType::from_str` is moved here from DataFusion.

# Are there any user-facing changes?

`DataType::from_str` is now public.

### CODE REVIEW

File: arrow-schema/src/lib.rs [hunk 1]
```diff
@@ -19,6 +19,7 @@
 
 mod datatype;
 pub use datatype::*;
+mod datatype_parse;
```
Comment (alamb) [inline arrow-schema/src/lib.rs:22]: It looks like datatpe_parse wasn't checked in perhaps

Comment (samuelcolvin) [reply arrow-schema/src/lib.rs:22]: 🤦 I was rushing, I'll add it when I'm back at my desk.

Comment (samuelcolvin) [reply arrow-schema/src/lib.rs:22]: added.

File: arrow-schema/src/datatype_parse.rs [hunk 2]
```diff
@@ -0,0 +1,787 @@
...
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::{fmt::Display, iter::Peekable, str::Chars, sync::Arc};
+
+use crate::{DataType, ArrowError, TimeUnit, IntervalUnit, Field};
+
+pub(crate) fn parse_data_type(val: &str) -> ArrowResult<DataType> {
```
Comment (samuelcolvin) [inline arrow-schema/src/datatype_parse.rs:22]: @alamb we could make this public and return `Result<DataType, String>`. So we can use it in datafusion `arrow_cast` without having to consider the other members of the `ArrowError` enum.

Or we could leave this as is and in `arrow_cast` do

```rs
match arrow_error_from_parse_data_type {
    ArrowError:ParseError(e) => DatafusinoError:Plan(e),
    e => DatafusionError:ArrowError(e, None),
}
```

LMK what you think.

Comment (alamb) [reply arrow-schema/src/datatype_parse.rs:22]: > Or we could leave this as is and in `arrow_cast` do

I think leaving the API surface area small (and hence not making this pub) is the better choice.

**Review by alamb — APPROVED** [review]
Looks great to me @samuelcolvin ❤️ 

I got inspired to improve the DataType documentation as well based on this PR (to make the feature potentially more discoverable -- PR in https://github.com/apache/arrow-rs/pull/5997)

**Comment by alamb** [comment]
In terms of releases, I think this one will likely appear in arrow 52.2.0:  https://github.com/apache/arrow-rs/issues/5998

https://github.com/apache/arrow-rs/issues/5905 was cut yesterday

**Comment by samuelcolvin** [comment]
That's fine, thanks for the quick response.

If it becomes a problem, I can copy and paste `datatype_parse.rs` into our code base for a few weeks - it's pretty well self-contained.

**Comment by alamb** [comment]
FWIW I made https://github.com/apache/datafusion/pull/11254 to use this function upstream in DataFusion ❤️

**Pull request merged**